redis/tests/test_cluster_async.rs

```diff
@@ -380,52 +380,60 @@ fn test_async_cluster_tryagain_exhaust_retries() {
 fn test_async_cluster_rebuild_with_extra_nodes() {
     let name = "rebuild_with_extra_nodes";
 
-    let requests = atomic::AtomicUsize::new(0);
+    let requests = Arc::new(AtomicI32::new(0));
     let started = atomic::AtomicBool::new(false);
     let MockEnv {
         runtime,
         async_connection: mut connection,
         handler: _handler,
         ..
-    } = MockEnv::new(name, move |cmd: &[u8], port| {
-        if !started.load(atomic::Ordering::SeqCst) {
-            respond_startup(name, cmd)?;
-        }
-        started.store(true, atomic::Ordering::SeqCst);
+    } = MockEnv::new(name, {
+        let i = requests.clone();
+        move |cmd: &[u8], port| {
+            if !started.load(atomic::Ordering::SeqCst) {
+                respond_startup_two_nodes(name, cmd)?;
+            }
+            started.store(true, atomic::Ordering::SeqCst);
 
-        if contains_slice(cmd, b"PING") {
-            return Err(Ok(Value::Status("OK".into())));
-        }
+            if contains_slice(cmd, b"PING") {
+                return Err(Ok(Value::Status("OK".into())));
+            }
 
-        let i = requests.fetch_add(1, atomic::Ordering::SeqCst);
-
-        match i {
-            // Respond that the key exists elswehere (the slot, 123, is unused in the
-            // implementation)
-            0 => Err(parse_redis_value(b"-MOVED 123\r\n")),
-            // Respond with the new masters
-            1 => Err(Ok(Value::Bulk(vec![
-                Value::Bulk(vec![
-                    Value::Int(0),
-                    Value::Int(1),
+            let i = i.fetch_add(1, atomic::Ordering::SeqCst);
+
+            match i {
+                // Respond that the key exists elswehere (the slot, 123, is unused in the
+                // implementation)
+                0 => Err(parse_redis_value(
+                    format!("-MOVED 14000 {name}:6380\r\n").as_bytes(),
+                )),
+                1 => {
```

Comment:
this might happen concurrently with the node refresh process, no? We're not promised that this will happen before the `CLUSTER SLOTS` command is sent.
I worry that the test will become flaky.

Reply:
Of course you have me doubting myself now :sweat_smile: , but I think this is a safe assumption. The request is added to `in_flight_requests` (rather than `pending_requests` as before) [after redirection](https://github.com/redis-rs/redis-rs/pull/857/files#diff-5ddfd8b2718c21504d43c50b0490bc22de4189208d97f5604a2297124e529577R664), and the refresh should not happen until after `in_flight_requests` have been drained.

Reply:
Ok, clearly I'm wrong, and furthermore I realize I've altered the original intent of this test. Anyway, a useful problem to run into, because it's made clear to me that following the redirect for `MOVED` rather than immediately rebuilding is probably not a good idea given our current impl.... if a new node has been added to the cluster, the current implementation will retry x number of times on a random connection rather than immediately rebuild the slots, which is suboptimal, and probably the most likely scenario if a `MOVED` error is actually received. I think I need to strip everything out but the ask implementation here :grimacing:

Reply:
😱 
If you want, I have a fix just for ASK rebased over https://github.com/redis-rs/redis-rs/pull/843, so if you prefer we can merge that and I'll open a PR.

Reply:
I've made another pass at this w/ the `MOVED` changes in place and tried to better validate expected behavior in new/existing tests.